File: linked_lists/reversing_a_linked_list.py
class ListNode:
    def __init__(self, val: int):
        self.val = val
        self.next = None

def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    prev = None
    curr = head

    while head:
        # Ensure we don't lose the next node
        next_node = curr.next

        # Reverse direction of the pointer
        curr.next = prev

        # Set current node to prev for next node
        prev = curr

        # Move on
        curr = next_node
    
    return prev

File: linked_lists/reversing_a_linked_list.java
class ListNode
{
    int val;
    ListNode next;
    ListNode (int val)
    {
        this.val = val;
    }
}

public ListNode reverseList(ListNode head)
{
    ListNode prev = null;
    ListNode curr = head;

    while (curr != null)
    {
        //! Ensure we don't lose the next node
        ListNode nextNode = curr.next;

        //! Reverse direction of pointer
        curr.next = prev;

        //! Set current node to prev for the next node
        prev = curr;

        //! Move on
        curr = nextNode;
    }

    return prev;
}